kratos/containers/sparse_graph.h

```diff
@@ -0,0 +1,394 @@
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+//class to construct and store a matrix graph. Can be used to construct efficiently a CSR matrix (or other sparse matrix types)
+
+/** This class is designed to store a matrix graph, aimed at the fast construction of other
+ * sparse matrix formats (particularly CSR)
+ * IMPORTANT NOTE: it is BY DESIGN NOT threadsafe! (a graph should be computed in each thread and then merged)
+*/
+
+class SparseGraph
+{
+public:
+    ///@name Type Definitions
+    ///@{
+    typedef std::size_t IndexType;
+    typedef std::map<IndexType, std::unordered_set<IndexType> > GraphType; //using a map since we need it ordered
+
+    /// Pointer definition of SparseGraph
+    KRATOS_CLASS_POINTER_DEFINITION(SparseGraph);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    SparseGraph(){}
+
+    /// Destructor.
+    virtual ~SparseGraph(){}
+
+    /// Assignment operator. TODO: decide if we do want to allow it
+    SparseGraph& operator=(SparseGraph const& rOther)=delete;
+    // {
+    //     this->AddEntries(rOther.GetGraph());
+    //     return *this;
+    // }
+
+    /// Copy constructor. TODO: we need it otherwise the sendrecv does not work...
+    ///but i don't know why :-(
+    SparseGraph(const SparseGraph& rOther)
+    {
+        this->AddEntries(rOther);
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+    void Clear()
+    {
+        mGraph.clear();
+    }
+
+    void AddEntry(const IndexType RowIndex, const IndexType ColIndex)
+    {
+        mGraph[RowIndex].insert(ColIndex);
+    }
+
+    template<class TContainerType>
+    void AddEntries(const IndexType RowIndex, const TContainerType& rColIndices)
+    {
+        mGraph[RowIndex].insert(rColIndices.begin(), rColIndices.end());
+    }
+
+    template<class TIteratorType>
+    void AddEntries(const IndexType RowIndex,
+                    const TIteratorType& rColBegin,
+                    const TIteratorType& rColEnd
+                    )
+    {
+        mGraph[RowIndex].insert(rColBegin, rColEnd);
+    }
+
+    //adds a square FEM matrix, identified by rIndices
+    template<class TContainerType>
+    void AddEntries(const TContainerType& rIndices)
+    {
+        for(auto I : rIndices)
+            mGraph[I].insert(rIndices.begin(), rIndices.end());
+    }
+
+    void AddEntries(const SparseGraph& rOtherGraph)
+    {
+        for(const auto& item: rOtherGraph.GetGraph())
+        {
+            AddEntries(item.first, item.second);
+        }
+    }
+
+    void Finalize()
+    {
+
+    }
+
+    const GraphType& GetGraph() const{
+        return mGraph;
+    }
+
+    IndexType ExportCSRArrays(
+        vector<IndexType>& rRowIndices,
+        vector<IndexType>& rColIndices
+    ) //TODO: this function should be imported in the CSR matrix interface, not here
```

Comment:
Agree on moving it. Meanwhile, I would template the arguments se we can use it with types other than std::vector